cmd/zoekt-sourcegraph-indexserver/cleanup.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/google/zoekt"
+)
+
+func cleanup(indexDir string, repos []string, now time.Time) {
+	trashDir := filepath.Join(indexDir, ".trash")
+	if err := os.MkdirAll(trashDir, 0755); err != nil {
+		log.Printf("failed to create trash dir: %v", err)
+	}
+
+	trash := getShards(trashDir)
+	index := getShards(indexDir)
+
+	// trash: Remove old shards and conflicts with index
+	minAge := now.Add(-24 * time.Hour)
+	for repo, shards := range trash {
+		old := false
+		for _, shard := range shards {
+			if shard.ModTime.Before(minAge) {
+				old = true
+			} else if shard.ModTime.After(now) {
+				debug.Printf("trashed shard %s has timestamp in the future, reseting to now", shard.Path)
+				_ = os.Chtimes(shard.Path, now, now)
+			}
+		}
+
+		if _, conflicts := index[repo]; !conflicts && !old {
+			continue
+		}
+
+		log.Printf("removing old shards from trash for %s", repo)
+		removeAll(shards)
+		delete(trash, repo)
+	}
+
+	// index: Move missing repos from trash into index
+	for _, repo := range repos {
+		// Delete from index so that index will only contain shards to be
+		// trashed.
+		delete(index, repo)
+
+		shards, ok := trash[repo]
+		if !ok {
+			continue
+		}
+
+		log.Printf("restoring shards from trash for %s", repo)
+		moveAll(indexDir, shards)
+	}
+
+	// index: Move non-existant repos into trash
+	for _, shards := range index {
+		// Best-effort touch. If touch fails, we will just remove from the
+		// trash sooner.
+		for _, shard := range shards {
+			_ = os.Chtimes(shard.Path, now, now)
+		}
+
+		moveAll(trashDir, shards)
+	}
+}
+
+type shard struct {
+	Repo    string
+	Path    string
+	ModTime time.Time
+}
+
+func getShards(dir string) map[string][]shard {
+	d, err := os.Open(dir)
+	if err != nil {
+		debug.Printf("failed to getShards: %s", dir)
+		return nil
+	}
+	defer d.Close()
+	names, _ := d.Readdirnames(-1)
+	sort.Strings(names)
+
+	shards := map[string][]shard{}
+	for _, n := range names {
+		path := filepath.Join(dir, n)
+		fi, err := os.Stat(path)
+		if err != nil {
+			debug.Printf("stat failed: %v", err)
+			continue
+		}
+		if fi.IsDir() {
+			continue
+		}
+
+		name, ok := shardRepoName(path)
+		if !ok {
+			debug.Printf("failed to read shard: %v", err)
+			continue
+		}
+
+		shards[name] = append(shards[name], shard{
+			Repo:    name,
+			Path:    path,
+			ModTime: fi.ModTime(),
+		})
+	}
+	return shards
+}
+
+func shardRepoName(path string) (string, bool) {
+	f, err := os.Open(path)
```

Comment:
Should we `debug.Printf` errors in this function, or return them and log them in the parent function?

Reply:
In the old implementation (where this was extracted from) we didn't do any logging. I'm happy to return an error instead and log it, since this shouldn't happen.